baybe/simulation/lookup.py

```diff
@@ -0,0 +1,209 @@
+        _lookup_targets_from_callable(queries, [t.name for t in targets], lookup)
+
+    # Get the results via dataframe lookup (works only for exact matches)
+    # IMPROVE: Although its not too important for a simulation, this
+    #  could also be implemented for approximate matches
+    elif isinstance(lookup, pd.DataFrame):
+        _lookup_targets_from_dataframe(queries, targets, lookup, impute_mode)
+
+
+def _lookup_targets_from_callable(
+    queries: pd.DataFrame,
+    target_names: Collection[str],
+    lookup: CallableLookup,
+) -> None:
+    """Look up target values via a callable."""
+    # Evaluate the callable
+    responses = lookup(queries)
+
+    # Assert that all targets are contained in the response
+    if not (exp := set(target_names)).issubset(act := set(responses)):
+        raise ValueError(
+            f"The provided lookup callable yielded values for the labels {act} but "
+            f"required are values for {exp}."
+        )
+
+    # Insert the target values in-place
+    queries[responses.columns] = responses
+
+
+def _lookup_targets_from_dataframe(
+    queries: pd.DataFrame,
+    targets: List[Target],
+    lookup: CallableLookup,
+    impute_mode: Literal["error", "worst", "best", "mean", "random"] = "error",
+) -> None:
+    """Look up target values from a dataframe."""
+    target_names = [t.name for t in targets]
+    all_match_vals = []
+
+    # IMPROVE: speed up the matching using a join
+    for _, row in queries.iterrows():
+        ind = lookup[
+            (lookup.loc[:, row.index] == row).all(axis=1, skipna=False)
+        ].index.values
+
+        if len(ind) > 1:
+            # More than one instance of this parameter combination
+            # have been measured
+            warnings.warn(
+                f"The lookup rows with indexes {ind} seem to be "
+                f"duplicates regarding parameter values. Choosing a "
+                f"random one.",
+                UserWarning,
+            )
+            match_vals = lookup.loc[np.random.choice(ind), target_names]
+
+        elif len(ind) < 1:
+            # Parameter combination cannot be looked up and needs to be
+            # imputed.
+            if impute_mode == "ignore":
+                raise AssertionError(
+                    "Something went wrong for impute_mode 'ignore'. "
+                    "It seems the search space was not correctly "
+                    "reduced before recommendations were generated."
+                )
+            match_vals = _impute_lookup(row, lookup, targets, impute_mode)
+
+        else:
+            # Exactly one match has been found
+            match_vals = lookup.loc[ind[0], target_names]
+
+        # Collect the matches
+        all_match_vals.append(match_vals)
+
+    # Add the lookup values
+    queries.loc[:, target_names] = pd.concat(all_match_vals, axis=1).T.values
+
+
+def _impute_lookup(
+    row: pd.Series,
+    lookup: pd.DataFrame,
+    targets: List[NumericalTarget],
```

Comment:
I am a bit unhappy that `targets` is something different in different functions: Here, it is a `List[NumericalTarget]`, in the other function it is a `List[Target]`. 
Is this really necessary? If so, could we maybe slightly adjust the names such that this is unique across function (e.g. by renaming it `numerical_targets` for this function)?

Reply:
Agree that this bad and still needs to be fixed, but don't worry for now --> that will be automatically resolved in the mypy PR for this module. The current PR only sets the stage for the docs. As mentioned in the PR description, there is still quite A LOT that should be fixed :D This module is very much one of our legacy pieces, containing quite a bit of technical debt ...

Reply:
Ok, then let's hope mypy will remind us of this issue